language/typescript-expression-plugin/src/service.ts

```diff
@@ -0,0 +1,360 @@
+    if (token?.type === 'Identifier') {
+      const expression = this._expressions.entries.get(token.name);
+
+      if (expression) {
+        const completionDetails = this.getCompletionEntryDetails(
+          context,
+          position,
+          expression.name
+        );
+
+        return {
+          ...completionDetails,
+          textSpan: toTSLocation(token),
+        };
+      }
+    }
+
+    return undefined;
+  }
+
+  private checkNode(
+    context: TemplateContext,
+    node: ExpressionNode,
+    xlrType: NodeType
+  ): ts.Diagnostic[] {
+    const diagnostics: ts.Diagnostic[] = [];
+    const asJsonNodeValue = convertExprToJSONNode(node);
+
+    if (asJsonNodeValue) {
+      const xlrDiags = this.xlr.validateByType(xlrType, asJsonNodeValue);
+
+      xlrDiags.forEach((d) => {
+        diagnostics.push({
+          file: context.node.getSourceFile(),
+          ...toTSLocation(node),
+          messageText: d.message,
+          category: ts.DiagnosticCategory.Error,
+          code: 1,
+        });
+      });
+    }
+
+    return diagnostics;
+  }
+
+  private getDiagnosticsForNode(
+    context: TemplateContext,
+    node: ExpressionNode
+  ): ts.Diagnostic[] {
+    const diags: ts.Diagnostic[] = [];
+
+    if (node.type === 'Compound') {
+      node.body.forEach((n) => {
+        diags.push(...this.getDiagnosticsForNode(context, n));
+      });
+    }
+
+    if (node.type === 'CallExpression') {
+      // Check that the expression is valid
+      const exprName = node.callTarget.name;
+      const expression = this._expressions.entries.get(exprName);
+
+      node.args.forEach((n) => {
+        diags.push(...this.getDiagnosticsForNode(context, n));
+      });
+
+      if (expression) {
+        // Double check the arguments match what we expect
+        const expectedArgs = expression.type.parameters;
+        const actualArgs = node.args;
+
+        actualArgs.forEach((actualArg, index) => {
+          const expectedArg = expectedArgs[index];
+
+          if (expectedArg) {
+            // Check that the type satisfies the expected type
+            diags.push(...this.checkNode(context, actualArg, expectedArg.type));
+          }
+        });
+
+        if (expectedArgs.length > actualArgs.length) {
+          const requiredArgs = expectedArgs.filter((a) => !a.optional);
+          if (actualArgs.length < requiredArgs.length) {
+            diags.push({
+              category: ts.DiagnosticCategory.Error,
+              code: 1,
+              file: context.node.getSourceFile(),
+              ...toTSLocation(node.callTarget),
+              messageText: `Expected ${requiredArgs.length} argument(s), got ${actualArgs.length}`,
+            });
+          }
+        }
+      } else {
+        diags.push({
+          category: ts.DiagnosticCategory.Error,
+          code: 1,
+          file: context.node.getSourceFile(),
+          ...toTSLocation(node.callTarget),
+          messageText: `Unknown expression ${exprName}`,
```

Comment:
Should we only push validations for expressions if there are plugins registered?